File: src/sdk/python/transport.py
"""Transport: spawn the server and speak the stream-json control protocol.

A :class:`Transport` owns the connection to the lite-harness server — process
spawn (or, for a fake, in-memory wiring), the single multiplexed NDJSON stream
described in ``PROTOCOL.md``, control-request/response correlation by
``request_id``, and delivery of the decoded non-control messages.

The wire language is the Claude Agent SDK stream-json control protocol:

* outgoing ``control_request`` lines (correlated by ``request_id``) for
  ``initialize`` / ``interrupt`` / ``set_permission_mode`` / ``set_model``,
* outgoing ``user`` lines to start a turn (no reply expected), and
* incoming lines demultiplexed on top-level ``type``: ``control_response`` lines
  resolve the matching pending request; everything else is a turn message.

:class:`SubprocessTransport` is the production implementation; the test fake
implements the same :class:`Transport` interface, so a fake server can be
injected wherever a real one would go.
"""

from __future__ import annotations

import asyncio
import json
import os
import secrets
import shlex
import shutil
from abc import ABC, abstractmethod
from pathlib import Path
from typing import Any, AsyncIterator, Callable

from errors import (
    CLIConnectionError,
    CLIJSONDecodeError,
    CLINotFoundError,
    ClaudeSDKError,
    ProcessError,
)

# Sentinel pushed onto the message queue when the stream ends.
_CLOSED = object()


class Transport(ABC):
    """Abstract stream-json control-protocol transport."""

    @abstractmethod
    async def connect(self) -> None:
        """Start the connection (spawn the process, begin reading)."""

    @abstractmethod
    async def send_control(self, subtype: str, **fields: Any) -> dict[str, Any]:
        """Send a ``control_request`` and await its matching ``control_response``.

        ``subtype`` is the request subtype (``initialize``, ``interrupt``,
        ``set_permission_mode``, ``set_model``); ``fields`` are merged into the
        ``request`` object. Returns the ``response`` dict on a ``success``
        subtype and raises a :class:`~lite_harness.errors.ClaudeSDKError`
        subclass on an ``error`` subtype or a transport failure.
        """

    @abstractmethod
    async def send_user_message(self, content: Any) -> None:
        """Write a ``user`` line to start a turn. No reply is expected."""

    @abstractmethod
    def messages(self) -> AsyncIterator[dict[str, Any]]:
        """Async iterator over incoming non-control message lines (raw dicts)."""

    @abstractmethod
    async def close(self) -> None:
        """Tear down the connection. Safe to call more than once."""


def resolve_server_command(explicit: list[str] | None = None) -> list[str]:
    """Resolve the base server spawn command per PROTOCOL.md.

    Order: explicit argument, then ``LITE_HARNESS_SERVER`` env var (a command
    line), then the in-repo server (when running from a clone), then the
    installed ``lite-harness-server`` on PATH. The stream-json launch flags are
    appended separately by :class:`SubprocessTransport`.
    """

    if explicit:
        return list(explicit)

    env_cmd = os.environ.get("LITE_HARNESS_SERVER")
    if env_cmd:
        return shlex.split(env_cmd)

    node = shutil.which("node")
    if node is None:
        raise CLINotFoundError(
            "Could not resolve the lite-harness server command. Set "
            "LITE_HARNESS_SERVER or pass a transport explicitly."
        )

    # Running straight from a clone: src/sdk/python -> src/sdk/server/server.mjs
    repo_server = Path(__file__).resolve().parent.parent / "server" / "server.mjs"
    if repo_server.is_file():
        return [node, str(repo_server)]

    return [node, "lite-harness-server"]


class SubprocessTransport(Transport):
    """Spawns the server and frames the stream-json control protocol over stdio."""

    def __init__(
        self,
        command: list[str] | None = None,
        *,
        agent: str | None = None,
        model: str | None = None,
        permission_mode: str | None = None,
        cwd: str | None = None,
        env: dict[str, str] | None = None,
        stderr: Callable[[str], None] | None = None,
    ) -> None:
        self._command = self._build_command(
            resolve_server_command(command),
            agent=agent,
            model=model,
            permission_mode=permission_mode,
            cwd=cwd,
        )
        self._cwd = cwd
        self._env = env
        self._stderr_cb = stderr

        self._proc: asyncio.subprocess.Process | None = None
        self._next_id = 0
        self._pending: dict[str, asyncio.Future[dict[str, Any]]] = {}
        self._messages: asyncio.Queue[Any] = asyncio.Queue()
        self._reader_task: asyncio.Task[None] | None = None
        self._stderr_task: asyncio.Task[None] | None = None
        self._closed = False

    @staticmethod
    def _build_command(
        base: list[str],
        *,
        agent: str | None,
        model: str | None,
        permission_mode: str | None,
        cwd: str | None,
    ) -> list[str]:
        cmd = [
            *base,
            "--input-format",
            "stream-json",
            "--output-format",
            "stream-json",
            "--verbose",
        ]
        if agent is not None:
            cmd += ["--agent", agent]
        if model is not None:
            cmd += ["--model", model]
        if permission_mode is not None:
            cmd += ["--permission-mode", permission_mode]
        if cwd is not None:
            cmd += ["--cwd", cwd]
        return cmd

    async def connect(self) -> None:
        spawn_env = {**os.environ, **(self._env or {})}
        try:
            self._proc = await asyncio.create_subprocess_exec(
                *self._command,
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                cwd=self._cwd,
                env=spawn_env,
            )
        except FileNotFoundError as exc:
            raise CLINotFoundError(
                f"Server command not found: {self._command[0]!r}"
            ) from exc
        except OSError as exc:
            raise CLIConnectionError(
                f"Failed to spawn server: {self._command!r}"
            ) from exc

        self._reader_task = asyncio.ensure_future(self._read_stdout())
        if self._proc.stderr is not None:
            self._stderr_task = asyncio.ensure_future(self._read_stderr())

    async def _read_stdout(self) -> None:
        assert self._proc is not None and self._proc.stdout is not None
        stdout = self._proc.stdout
        try:
            while True:
                raw = await stdout.readline()
                if not raw:
                    break
                line = raw.decode("utf-8").strip()
                if not line:
                    continue
                self._dispatch_line(line)
        except asyncio.CancelledError:
            raise
        finally:
            await self._fail_pending_and_close()

    def _dispatch_line(self, line: str) -> None:
        try:
            obj = json.loads(line)
        except json.JSONDecodeError as exc:
            error = CLIJSONDecodeError(line, exc)
            self._fail_all_pending(error)
            return

        if not isinstance(obj, dict):
            return

        if obj.get("type") == "control_response":
            response = obj.get("response") or {}
            request_id = response.get("request_id")
            if not isinstance(request_id, str):
                return
            future = self._pending.pop(request_id, None)
            if future is None or future.done():
                return
            if response.get("subtype") == "error":
                future.set_exception(
                    ClaudeSDKError(
                        f"Control request failed: {response.get('error')}"
                    )
                )
            else:
                future.set_result(response)
            return

        # Any non-control line is a turn message.
        self._messages.put_nowait(obj)

    async def _read_stderr(self) -> None:
        assert self._proc is not None and self._proc.stderr is not None
        stderr = self._proc.stderr
        try:
            while True:
                raw = await stderr.readline()
                if not raw:
                    break
                if self._stderr_cb is not None:
                    self._stderr_cb(raw.decode("utf-8").rstrip("\n"))
        except asyncio.CancelledError:
            raise

    def _fail_all_pending(self, error: Exception) -> None:
        for future in self._pending.values():
            if not future.done():
                future.set_exception(error)
        self._pending.clear()

    async def _fail_pending_and_close(self) -> None:
        if self._pending:
            stderr_text = await self._collect_stderr()
            exit_code = self._proc.returncode if self._proc is not None else None
            self._fail_all_pending(
                ProcessError(
                    "Server closed the connection before replying",
                    exit_code=exit_code,
                    stderr=stderr_text,
                )
            )
        self._messages.put_nowait(_CLOSED)

    async def _collect_stderr(self) -> str | None:
        if self._proc is None or self._proc.stderr is None:
            return None
        try:
            data = await self._proc.stderr.read()
        except Exception:
            return None
        text = data.decode("utf-8", errors="replace").strip()
        return text or None

    async def send_control(self, subtype: str, **fields: Any) -> dict[str, Any]:
        if self._proc is None or self._proc.stdin is None:
            raise CLIConnectionError("Transport is not connected")

        self._next_id += 1
        request_id = f"req_{self._next_id}_{secrets.token_hex(4)}"
        future: asyncio.Future[dict[str, Any]] = (
            asyncio.get_event_loop().create_future()
        )
        self._pending[request_id] = future

        payload = {
            "type": "control_request",
            "request_id": request_id,
            "request": {"subtype": subtype, **fields},
        }
        try:
            self._write_line(payload)
        except Exception:
            self._pending.pop(request_id, None)
            raise
        return await future

    async def send_user_message(self, content: Any) -> None:
        if self._proc is None or self._proc.stdin is None:
            raise CLIConnectionError("Transport is not connected")
        self._write_line(
            {
                "type": "user",
                "message": {"role": "user", "content": content},
                "session_id": None,
                "parent_tool_use_id": None,
            }
        )

    def _write_line(self, payload: dict[str, Any]) -> None:
        assert self._proc is not None and self._proc.stdin is not None
        line = json.dumps(payload, separators=(",", ":")) + "\n"
        self._proc.stdin.write(line.encode("utf-8"))

    async def messages(self) -> AsyncIterator[dict[str, Any]]:
        while True:
            item = await self._messages.get()
            if item is _CLOSED:
                return
            yield item

    async def close(self) -> None:
        if self._closed:
            return
        self._closed = True

        proc = self._proc
        if proc is not None:
            if proc.stdin is not None and not proc.stdin.is_closing():
                try:
                    proc.stdin.close()
                except Exception:
                    pass
            if proc.returncode is None:
                try:
                    proc.terminate()
                except ProcessLookupError:
                    pass
                try:
                    await asyncio.wait_for(proc.wait(), timeout=5.0)
                except asyncio.TimeoutError:
                    try:
                        proc.kill()
                    except ProcessLookupError:
                        pass
                    await proc.wait()
            else:
                # Already exited: still await so the loop runs the
                # connection-lost callback now, instead of letting the
                # subprocess transport's __del__ try to close pipes after the
                # loop is gone ("Event loop is closed" warning at GC time).
                await proc.wait()

        for task in (self._reader_task, self._stderr_task):
            if task is not None and not task.done():
                task.cancel()
                try:
                    await task
                except (asyncio.CancelledError, Exception):
                    pass

        self._messages.put_nowait(_CLOSED)


__all__ = ["Transport", "SubprocessTransport", "resolve_server_command"]

File: src/sdk/typescript/transport.ts
/**
 * Transport: the single place that owns process spawn, NDJSON framing over
 * stdio, the multiplexed Claude Agent SDK stream-json control protocol, and
 * routing of incoming lines into either a control-request correlation map
 * (keyed by `request_id`) or an internal pushable queue of decoded messages.
 *
 * See PROTOCOL.md. No agent/business logic lives here.
 *
 * One NDJSON stream carries everything. We demux every incoming line on its
 * top-level `type`:
 *   - `control_response`  -> resolve/reject the pending control request keyed
 *                            by `response.request_id`.
 *   - everything else     -> push the raw object into the message queue, where
 *                            `query.ts` decodes it into an `SDKMessage`.
 */

import { spawn, type ChildProcessWithoutNullStreams } from "node:child_process";
import { createInterface, type Interface } from "node:readline";
import { randomBytes } from "node:crypto";

import {
  AbortError,
  CLIConnectionError,
  CLINotFoundError,
  ProcessError,
} from "./errors.js";

/** Grace period (ms) to let the server exit on its own after stdin is ended. */
const GRACEFUL_EXIT_MS = 2000;

interface PendingRequest {
  resolve: (response: Record<string, unknown>) => void;
  reject: (error: Error) => void;
}

/** A single buffered incoming (non-control) wire object. */
export type IncomingLine = unknown;

/** How the transport should locate and launch the server process. */
export interface TransportConfig {
  /** Explicit command override (tests inject a fake server here). Highest priority. */
  command?: string[];
  /** Working directory for the spawned process. */
  cwd?: string;
  /** Environment variables for the spawned process. */
  env?: Record<string, string | undefined>;
  /** Receives server stderr line-by-line. */
  stderr?: (data: string) => void;
  /** Aborts the connection / kills the process. */
  abortController?: AbortController;
  /** Extra args appended to the spawn command (the stream-json flags etc.). */
  args?: string[];
}

/**
 * Resolve the server spawn command per PROTOCOL.md "Server command resolution":
 *   1. explicit transport argument,
 *   2. `LITE_HARNESS_SERVER` env var (a command line), else
 *   3. the bundled default.
 */
export function resolveServerCommand(
  explicit: string[] | undefined,
  env: Record<string, string | undefined>,
): string[] {
  if (explicit && explicit.length > 0) {
    return explicit;
  }
  const fromEnv = env.LITE_HARNESS_SERVER;
  if (fromEnv && fromEnv.trim().length > 0) {
    // A simple whitespace split is sufficient for the documented "command line".
    return fromEnv.trim().split(/\s+/);
  }
  // Bundled default. There is no packaged server in this scaffold, so we point
  // at a conventional location; callers without one should set the env var or
  // pass an explicit command.
  return ["node", "lite-harness-server"];
}

/**
 * Owns the child process and the NDJSON wire. Construct, then `sendControl()`
 * for control_request/control_response round-trips, `sendUserMessage()` to
 * start a turn, and pull buffered incoming lines with `nextLine()`.
 */
export class Transport {
  private child: ChildProcessWithoutNullStreams | null = null;
  private rl: Interface | null = null;
  private stderrRl: Interface | null = null;
  private graceTimer: ReturnType<typeof setTimeout> | null = null;
  private requestCounter = 0;
  private readonly pending = new Map<string, PendingRequest>();

  // Incoming-line pushable: a queue plus a single waiter.
  private readonly lineQueue: IncomingLine[] = [];
  private lineWaiter: ((value: IteratorResult<IncomingLine>) => void) | null = null;
  private linesDone = false;

  private closed = false;
  private connectError: Error | null = null;
  private stderrBuffer = "";

  constructor(private readonly config: TransportConfig) {}

  /** Spawn the server process and begin reading its stdout. */
  connect(): void {
    if (this.child) {
      return;
    }
    // Merge the caller-supplied env ON TOP of the current process env so a
    // partial `env` option never drops PATH and friends. Server command
    // resolution sees the merged view too (so a caller can set
    // LITE_HARNESS_SERVER without restating the whole environment).
    const env: Record<string, string | undefined> = {
      ...process.env,
      ...this.config.env,
    };
    const command = resolveServerCommand(this.config.command, env);
    const [bin, ...baseArgs] = command;
    if (!bin) {
      throw new CLINotFoundError("No server command could be resolved.");
    }
    const args = [...baseArgs, ...(this.config.args ?? [])];

    // Filter undefined env values for spawn's stricter typing.
    const spawnEnv: Record<string, string> = {};
    for (const [key, value] of Object.entries(env)) {
      if (typeof value === "string") {
        spawnEnv[key] = value;
      }
    }

    let child: ChildProcessWithoutNullStreams;
    try {
      child = spawn(bin, args, {
        cwd: this.config.cwd,
        env: spawnEnv,
        stdio: ["pipe", "pipe", "pipe"],
      });
    } catch (err) {
      throw new CLIConnectionError(`Failed to spawn server "${bin}": ${String(err)}`);
    }
    this.child = child;

    child.on("error", (err: NodeJS.ErrnoException) => {
      const error =
        err.code === "ENOENT"
          ? new CLINotFoundError(`Server command not found: "${bin}"`)
          : new CLIConnectionError(`Server process error: ${err.message}`);
      this.failAll(error);
    });

    child.on("exit", (code) => {
      if (this.closed) {
        this.finishLines();
        return;
      }
      // Unexpected exit while requests are outstanding.
      const error = new ProcessError("Server process exited", {
        exitCode: code,
        stderr: this.stderrBuffer || null,
      });
      this.failAll(error);
    });

    this.stderrRl = createInterface({ input: child.stderr });
    this.stderrRl.on("line", (line) => {
      this.stderrBuffer += line + "\n";
      this.config.stderr?.(line);
    });

    this.rl = createInterface({ input: child.stdout });
    this.rl.on("line", (line) => this.handleLine(line));

    const abort = this.config.abortController;
    if (abort) {
      if (abort.signal.aborted) {
        this.kill();
      } else {
        abort.signal.addEventListener("abort", () => this.kill(), { once: true });
      }
    }
  }

  /**
   * Reject every still-pending control-request promise so a bare
   * `await q.interrupt()` racing teardown never hangs forever. Used on abort /
   * kill. Defaults to {@link AbortError}.
   */
  private rejectPending(error: Error): void {
    for (const pending of this.pending.values()) {
      pending.reject(error);
    }
    this.pending.clear();
  }

  private handleLine(line: string): void {
    const trimmed = line.trim();
    if (!trimmed) {
      return;
    }
    let parsed: unknown;
    try {
      parsed = JSON.parse(trimmed);
    } catch {
      // Forward-compat (PROTOCOL.md): a single unparseable stdout line must NOT
      // tear down the session. Skip it — optionally surfacing it as a stderr
      // diagnostic — and keep reading. A newer server emitting a line shape an
      // older SDK can't parse should never kill an in-flight turn.
      this.config.stderr?.(`[lite-harness] skipped unparseable stdout line: ${trimmed}`);
      return;
    }
    if (typeof parsed !== "object" || parsed === null) {
      return;
    }
    const obj = parsed as Record<string, unknown>;

    // Control response: route to the matching pending control request.
    if (obj.type === "control_response") {
      this.handleControlResponse(obj);
      return;
    }

    // Everything else is a turn message — push it to the queue for decoding.
    this.pushLine(obj);
  }

  private handleControlResponse(obj: Record<string, unknown>): void {
    const response =
      typeof obj.response === "object" && obj.response !== null
        ? (obj.response as Record<string, unknown>)
        : {};
    const requestId = typeof response.request_id === "string" ? response.request_id : null;
    if (requestId === null) {
      return;
    }
    const pending = this.pending.get(requestId);
    if (!pending) {
      return;
    }
    this.pending.delete(requestId);
    if (response.subtype === "error") {
      const message =
        typeof response.error === "string" ? response.error : "control request failed";
      pending.reject(new CLIConnectionError(`Server control error: ${message}`));
    } else {
      pending.resolve(response);
    }
  }

  private newRequestId(): string {
    const suffix = randomBytes(3).toString("hex");
    return `req_${++this.requestCounter}_${suffix}`;
  }

  /**
   * Send a `control_request` with the given subtype + extra fields and await the
   * matching `control_response`. Rejects if the server replies with an error
   * subtype.
   */
  sendControl(
    subtype: string,
    fields: Record<string, unknown> = {},
  ): Promise<Record<string, unknown>> {
    if (this.connectError) {
      return Promise.reject(this.connectError);
    }
    if (!this.child || this.closed) {
      return Promise.reject(new CLIConnectionError("Transport is not connected."));
    }
    const requestId = this.newRequestId();
    const payload =
      JSON.stringify({
        type: "control_request",
        request_id: requestId,
        request: { subtype, ...fields },
      }) + "\n";
    return new Promise<Record<string, unknown>>((resolve, reject) => {
      this.pending.set(requestId, { resolve, reject });
      this.write(payload, (err) => {
        if (err) {
          this.pending.delete(requestId);
          reject(err);
        }
      });
    });
  }

  /**
   * Start a turn by writing a `user` line. There is no request/response — the
   * server simply streams system/assistant/result lines until the turn ends.
   */
  sendUserMessage(content: string | unknown[]): Promise<void> {
    if (this.connectError) {
      return Promise.reject(this.connectError);
    }
    if (!this.child || this.closed) {
      return Promise.reject(new CLIConnectionError("Transport is not connected."));
    }
    const payload =
      JSON.stringify({
        type: "user",
        message: { role: "user", content },
        session_id: null,
        parent_tool_use_id: null,
      }) + "\n";
    return new Promise<void>((resolve, reject) => {
      this.write(payload, (err) => {
        if (err) {
          reject(err);
        } else {
          resolve();
        }
      });
    });
  }

  private write(payload: string, done: (err?: Error) => void): void {
    try {
      this.child!.stdin.write(payload, (err) => {
        if (err) {
          done(new CLIConnectionError(`Failed to write to server: ${err.message}`));
        } else {
          done();
        }
      });
    } catch (err) {
      done(new CLIConnectionError(`Failed to write to server: ${String(err)}`));
    }
  }

  /** Pull the next buffered incoming line, awaiting one if the queue is empty. */
  nextLine(): Promise<IteratorResult<IncomingLine>> {
    if (this.lineQueue.length > 0) {
      const value = this.lineQueue.shift() as IncomingLine;
      return Promise.resolve({ value, done: false });
    }
    if (this.linesDone) {
      return Promise.resolve({ value: undefined, done: true });
    }
    return new Promise((resolve) => {
      this.lineWaiter = resolve;
    });
  }

  private pushLine(line: IncomingLine): void {
    if (this.lineWaiter) {
      const waiter = this.lineWaiter;
      this.lineWaiter = null;
      waiter({ value: line, done: false });
    } else {
      this.lineQueue.push(line);
    }
  }

  private finishLines(): void {
    this.linesDone = true;
    if (this.lineWaiter) {
      const waiter = this.lineWaiter;
      this.lineWaiter = null;
      waiter({ value: undefined, done: true });
    }
  }

  private failAll(error: Error): void {
    this.connectError = error;
    for (const pending of this.pending.values()) {
      pending.reject(error);
    }
    this.pending.clear();
    this.finishLines();
  }

  /**
   * Immediately terminate the child process and stop reading. Reserved for
   * abort / early `return()` / error paths — see {@link shutdownGraceful} for
   * normal completion. Rejects any still-pending control promises so callers
   * racing teardown never hang.
   */
  kill(): void {
    this.closed = true;
    if (this.graceTimer) {
      clearTimeout(this.graceTimer);
      this.graceTimer = null;
    }
    this.rl?.close();
    this.rl = null;
    this.stderrRl?.close();
    this.stderrRl = null;
    if (this.child) {
      try {
        this.child.stdin.end();
      } catch {
        // ignore
      }
      this.child.kill();
      this.child = null;
    }
    this.rejectPending(new AbortError("Transport closed before control response."));
    this.finishLines();
  }

  /**
   * Graceful shutdown for NORMAL completion (PROTOCOL.md: a one-shot turn ends
   * stdin and lets the server exit on its own). End stdin, wait a short grace
   * period for the process to exit, then fall back to {@link kill} if it has
   * not. Resolves once the process is gone (or killed).
   */
  shutdownGraceful(): Promise<void> {
    this.closed = true;
    const child = this.child;
    if (!child) {
      this.kill();
      return Promise.resolve();
    }
    return new Promise<void>((resolve) => {
      const finalize = (): void => {
        if (this.graceTimer) {
          clearTimeout(this.graceTimer);
          this.graceTimer = null;
        }
        // kill() is idempotent and closes readers + rejects any leftover
        // pending promises; safe to call even after a clean exit.
        this.kill();
        resolve();
      };
      child.once("exit", finalize);
      try {
        child.stdin.end();
      } catch {
        // ignore
      }
      this.graceTimer = setTimeout(finalize, GRACEFUL_EXIT_MS);
      // Don't keep the event loop alive purely for the grace timer.
      this.graceTimer.unref?.();
    });
  }

  /** Mark closed (after a graceful shutdown) so an exit is not treated as an error. */
  markClosed(): void {
    this.closed = true;
  }
}
